### ISSUE TITLE
Why AUC ~0.5 using pointwise-based sigmoid_cross_entropy_loss?

### ISSUE THREAD

**Issue by LiangqunLu** [tensorflow/ranking#172]
Hi,

When I used listwise-based softmax loss for TF Ranking, I got fair NDCG@, ～0.66, however the AUC was like 0.5-0.52. This is frustrating. This may be caused by the discrepancy between classification and listwise ranking. However, when I changed to sigmoid cross entropy loss, I got AUC 0.5, ~0.67 NDCG@3. Well, this is beyond understanding. I thought the sigmoid cross entropy loss is supposed to optimize the accuracy in the classification problem. There should be a better AUC in the end. Could somebody please point out my mistake here or explain the difference of TF-Ranking pointwise classification and the normal sklearn-based classification? Thanks!

**Comment by lyyanlely** [comment]
@LiangqunLu , Can you please clarify a bit your problem? My understanding is that you are working on a classification problem and tried using sigmoid cross-entropy loss and softmax loss. Theoretically, the sigmoid cross-entropy loss would be exactly what you use for a two-class classification problem. In general, you may want to check that you are not mixing logits with probabilities (sigmoids of logits) when you pass them to the loss function and AUC metric. If that's not the problem, can you please share your training with tensorboard or something similar? How did you compute the metrics?

**Comment by LiangqunLu** [comment]
Sure. Sorry for being confused. I am working on a Ranking problem, in which I would like to evaluate mainly 2 metrics, NDCG@3 and AUC. The former is highly related to ranking while the latter is for the basic classification performance. I tried different strategies, you know, to have a good performance eventually. My data has queries and corresponding various numbers of documents. For the labels, I am using binary at this stage, meaning {0, 1}. 

Before I ran my code, I added AUC metric as here #168. The NDCG@3 is in the example script "tf_ranking_libsvm.py". 

At first, I used listwise-based softmax loss for training with group size default 1. In the end, the curves for ndcg and auc are below. The two lines in each plot indicate 2 different datasets I used. The better ndcg@3 is 0.67, however auc is 0.49 - 0.51. I feel somehow AUC should be better like 0.6 given ndcg@3 is improving.

// NDCG@3
![image](https://user-images.githubusercontent.com/10750332/75500607-54b52680-5982-11ea-96d9-03dfbd3d4239.png)

// AUC
![image](https://user-images.githubusercontent.com/10750332/75500629-68f92380-5982-11ea-8b81-550a8befc10f.png)

I next tested pointwise based sigmoid_cross_entropy_loss in the hope that TFR is working as a binary classifier. However, the AUC stays like 0.5, not getting better. Well, I expected a better AUC in the classification task. So I am a little confused.

**Comment by lyyanlely** [comment]
Hi Liangqun, your training curves do not appear to converge to me. Can you please also share your loss curve to see if that is decreasing? The ndcg@3 metric around 0.6 doesn't mean your model behaves correctly. If, for example, you have a list of size 3 with only one positive example, random order of the list will still give you a good ndcg@3 around 0.7. Your AUC fluctuates around 0.5 is a clear sign of not training.

**Comment by lyyanlely** [comment]
Hi Lianqun, has my last reply helped on your confusion? Please let me know if you need further help. Otherwise, I'll close the issue.

**Comment by LiangqunLu** [comment]
Hi @lyyanlely , thanks for checking out. Your previous comment is very helpful. But I have another serious problem about TF-Ranking library. I just pasted images below. 

// Loss
![image](https://user-images.githubusercontent.com/10750332/76365597-e3ac3200-62e4-11ea-921a-5748ce471e7f.png)

// AUC
![image](https://user-images.githubusercontent.com/10750332/76365612-ec9d0380-62e4-11ea-8b51-627dc588f1f3.png)

// MAP@3 and MAP@5, red line

![image](https://user-images.githubusercontent.com/10750332/76365662-1a824800-62e5-11ea-9360-6792cc08719c.png)

//ndcg@3 and ndcg@5, red line

![image](https://user-images.githubusercontent.com/10750332/76365717-4271ab80-62e5-11ea-94d6-05caf170ac3f.png)

I have used TF-Ranking for multiple data sets in one task. In the one dataset, the classes are imbalanced and the number of query documents vary. From the images, the loss curves are not converged well, and AUC is alright to be around 0.6, though it keeps decreasing. But the MAP and NDCG are like extremely low, around 0.05. In other datasets, although the AUCs are not pretty high, but the NDCG and MAP values are like 0.6+. 

I may have missed some important details. Do you have explanations for the possible reasons? Thanks!

**Comment by lyyanlely** [comment]
Hi @LiangqunLu, Could you please explain a bit? I can see your model is trained with the orange line decreasing. Which dataset did you train on? Is it the one same as the pink lines? From all other metrics (expect for the orange loss), your model does not appear to improve at all. There are many possibilities. 

My general suggestion without many details about your datasets would be to try your model on a commonly used baseline dataset see if it can learn (with loss decreasing and metrics (AUC and NDCG) increasing). If it does, you can then check on your datasets. Don't rely too much on the absolute values of NDCG or MAP. The different nature of two datasets could lead to very different absolute values.

**Comment by LiangqunLu** [comment]
For the loss in the orange line (loss images), that is the training data and the red line indicates the validation data. The pink line in the MAP and NDCG metrics is from another validation dataset, quite similar with the red line data set. That is why I did not figure out the huge difference in the NDCG metrics. The loss curve is decreasing maybe a little bit, but the AUC is decreasing. I may explore the difference between the 2 data sets. Also I include other scripts to have some basic tests.

Could you help me check out one of my new and urgent issues listed as #174 , please?

**Comment by lyyanlely** [comment]
@LiangqunLu Xuanhui has replied your issue last week. Hope your issue has been resolved there. For this issue, have you been able to make it work now? In terms of metrics (AUC, MAP, and NDCG), I would say your model didn't learn successfully from the training set, either. If your red validation set is a subsampling of the training set, you should have metrics on training (orange) behave similar to the red lines, which you didn't show. 
I could help better if you can share your estimator and training pipeline with some example fake data.

**Comment by LiangqunLu** [comment]
Hi @lyyanlely , Thanks for checking out. I followed your advice back then, and I got improved performance, well at least more reasonable curves.

![image](https://user-images.githubusercontent.com/10750332/78838131-596df100-79aa-11ea-92d9-c9f62ec52ed4.png)

The orange line represents training loss while the blue curve is the loss for validation data at each 1000 steps. There are decreasing training loss, slightly decreasing validation loss and responding AUC curve. I feel the training is working on my data set. 

For another validation data set with a different class ratio which a majority of queries have 1 document or have all relevance 0 documents.

![image](https://user-images.githubusercontent.com/10750332/78838594-63dcba80-79ab-11ea-864e-d78cdbde6783.png)

The extremely low MAP@3 and NDCG@3 really have worried me. I think I am fine with slowly increasing AUC. But it is difficult for me to explain the MAP@3. 

By the way, did you guys have the TF-Ranking workshop?
